plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,440 @@
+                    pos,
+                    "E_x",
+                    "E_y",
+                    "E_z",
+                    "B_x",
+                    "B_y",
+                    "B_z",
+                    persistent=True,
+                )
+
+            # Interpret any NaN values (points off the grid) as zero
+            # Do this before adding to the totals, because 0 + nan = nan
+            _Ex = np.nan_to_num(_Ex, nan=0.0 * u.V / u.m)
+            _Ey = np.nan_to_num(_Ey, nan=0.0 * u.V / u.m)
+            _Ez = np.nan_to_num(_Ez, nan=0.0 * u.V / u.m)
+            _Bx = np.nan_to_num(_Bx, nan=0.0 * u.T)
+            _By = np.nan_to_num(_By, nan=0.0 * u.T)
+            _Bz = np.nan_to_num(_Bz, nan=0.0 * u.T)
+
+            # Add the values interpolated for this grid to the totals
+            Ex += _Ex
+            Ey += _Ey
+            Ez += _Ez
+            Bx += _Bx
+            By += _By
+            Bz += _Bz
+
+        # Create arrays of E and B as required by push algorithm
+        E = np.array(
+            [Ex.to(u.V / u.m).value, Ey.to(u.V / u.m).value, Ez.to(u.V / u.m).value]
+        )
+        E = np.moveaxis(E, 0, -1)
+        B = np.array([Bx.to(u.T).value, By.to(u.T).value, Bz.to(u.T).value])
+        B = np.moveaxis(B, 0, -1)
+
+        # Calculate the adaptive timestep from the fields currently experienced
+        # by the particles
+        # If user sets dt explicitly, that's handled in _adaptive_dt
+        dt = self._adaptive_dt(Ex, Ey, Ez, Bx, By, Bz)
+
+        # TODO: Test v/c and implement relativistic Boris push when required
+        # vc = np.max(v)/_c
+
+        # If dt is not a scalar, make sure it can be multiplied by an
+        # [nparticles, 3] shape field array
+        if dt.size > 1:
+            dt = dt[:, np.newaxis]
+
+        self.x, self.v = boris_push(
+            self.x, self.v, B, E, self.q, self.m, dt, inplace=False
+        )
+
+    @property
+    def on_any_grid(self):
+        """
+        Binary array for each particle indicating whether it is currently
+        on ANY grid.
+        """
+        return np.sum(self.on_grid, axis=-1) > 0
+
+    @staticmethod
+    def _stop_condition():
+        return False
+
+    @property
+    def vmax(self):
+        """
+        Calculate the maximum velocity.
+        Used for determining the grid crossing maximum timestep.
+        """
+        return np.max(np.linalg.norm(self.v, axis=-1))
+
+    def _validate_inputs(self, field_weighting: str):
+        # Load and validate inputs
+        field_weightings = ["volume averaged", "nearest neighbor"]
+        if field_weighting in field_weightings:
+            self.field_weighting = field_weighting
+        else:
+            raise ValueError(
+                f"{field_weighting} is not a valid option for ",
+                "field_weighting. Valid choices are",
+                f"{field_weightings}",
+            )
+
+        # Check to make sure particles have already been generated
```

Comment:
Either way, this should definitely just be in run

Reply:
The `test_run_options` test doesn't like it when I do the validation in the general `run` function so I had to make sure this condition is checked in both `run` functions

Reply:
I've moved this to the `_enforce_particle_creation` method which seems like a better spot for it